### PR TITLE
[package-config.cmake] Handle in a friendly user manner optional package

### PR DESCRIPTION
**Pull request by olivier-stasse**
An optional package is specified in the following cases:
ADD_PROJECT_DEPENDENCY(package)
ADD_PROJECT_DEPENDENCY(package OPTIONAL)
In this case if the package is not detected and this is not put in the dependency list.

### CODE REVIEW

**Comment by nim65s** [comment]
Hi,

`ADD_PROJECT_DEPENDENCY(package)` is already optional ; because if it is not, you should use
`ADD_PROJECT_DEPENDENCY(package REQUIRED)`.

Moreover, this macro mimics the API of the upstream ones it calls: `find_package` / `find_dependency`, so I think we should try to stick to this API as much as possible, unless there an issue you want to solve ?

**Comment by jmirabel** [comment]
1. is it an issue to have an `find_package(Foo)` in the BarConfig.cmake ? Even if, for aliasing reason, Foo wasn't found at compile time but is found by packages depending on Bar, the targets from Foo will not be used by Bar.
2. to avoid adding optional dependencies to list of dependencies, you should merely add them after the `find_package`, only if `package_FOUND` is true. It will cover all cases, with no change in the API and a very minimal change in the code.

**Comment by olivier-stasse** [comment]
* I am fine with not putting OPTIONAL (i just find it more clear to the user)
 * If I am not mistaken right now, I have to do:
```
find_package(gazebo)
if (GAZEBO_FOUND)
  ADD_PROJECT_DEPENDENCY(gazebo)
endif(GAZEBO_FOUND)
```
the proposed PR simply replace those lines by:
```
ADD_PROJECT_DEPENDENCY(gazebo)
```
if REQUIRED is not specified then it is optional.
Does that make sense ?

**Comment by jmirabel** [comment]
>     * If I am not mistaken right now, I have to do:
> 
> 
> ```
> find_package(gazebo)
> if (GAZEBO_FOUND)
>   ADD_PROJECT_DEPENDENCY(gazebo)
> endif(GAZEBO_FOUND)
> ```

Why do you have to do this ?

**Comment by olivier-stasse** [comment]
As far as I tested ADD_PROJECT_DEPENDENCY will add Gazebo to the list of dependency even if it is not present.
So you need to quietly test it before, and then add it if it is detected.
See https://github.com/jrl-umi3218/jrl-cmakemodules/issues/391 on how Pierre is handling optional package.

File: package-config.cmake [hunk 1]
```diff
@@ -70,21 +70,49 @@ MACRO(ADD_PROJECT_DEPENDENCY)
...
-  string(REPLACE ";" " " PACKAGE_ARGS "${PARSED_ARGN_UNPARSED_ARGUMENTS}")
-  list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_FIND_PACKAGE "  find_package(${PACKAGE_ARGS})")
-  list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_FIND_DEPENDENCY "  find_dependency(${PACKAGE_ARGS})")
-  find_package(${PARSED_ARGN_UNPARSED_ARGUMENTS})
+  string(REPLACE ";" " " PACKAGE_ARGS
+    "${PARSED_ARGN_UNPARSED_ARGUMENTS}")
+
+  # If REQUIRED is not specified consider the package as optional
+  # search for the package in a quiet mode.
+  IF(PARSED_ARGN_OPTIONAL OR NOT PARSED_ARGN_REQUIRED)
```
Comment (gergondet) [inline package-config.cmake:92]: Please avoid:
- using all caps IF/ELSE/ENDIF
- using the matching arguments in else and endif conditions, they have not been required for a really long time

File: package-config.cmake [hunk 2]
```diff
@@ -70,21 +70,49 @@ MACRO(ADD_PROJECT_DEPENDENCY)
...
+    find_package(${PARSED_ARGN_UNPARSED_ARGUMENTS} REQUIRED)
+  ENDIF(PARSED_ARGN_OPTIONAL OR NOT PARSED_ARGN_REQUIRED)
+
+  # If the package has been found provides the dependency
+  IF(${${ARGV0}_FOUND})
+
+    list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_FIND_PACKAGE
+      "  find_package(${PACKAGE_ARGS})")
+    list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_FIND_DEPENDENCY
+      "  find_dependency(${PACKAGE_ARGS})")
```
Comment (gergondet) [inline package-config.cmake:105]: The generated file will miss the `REQUIRED` argument since you consume it in `cmake_parse_arguments`

**Comment by jmirabel** [comment]
First of all, the issue you mention is purely cosmetic. The only difference between
```cmake
find_package(gazebo)
if (GAZEBO_FOUND)
  ADD_PROJECT_DEPENDENCY(gazebo)
endif(GAZEBO_FOUND)
```
and
```cmake
ADD_PROJECT_DEPENDENCY(gazebo)
```
is that, in the second case, the `Config.cmake` will look for gazebo (without failing if not found) and will never use this dependency.

**Comment by jmirabel** [comment]
So both version should have the same behavior.

Now, if you really want to solve the cosmetic issue, I think the API does not have to be modified. You can simply do
```cmake

MACRO(ADD_PROJECT_DEPENDENCY)
  # add dependency to the generated .pc
  # ref https://github.com/jrl-umi3218/jrl-cmakemodules/pull/335
  cmake_parse_arguments(PARSED_ARGN "" "PKG_CONFIG_REQUIRES" "" ${ARGN})
  IF(PARSED_ARGN_PKG_CONFIG_REQUIRES)
    _ADD_TO_LIST_IF_NOT_PRESENT(_PKG_CONFIG_REQUIRES "${PARSED_ARGN_PKG_CONFIG_REQUIRES}")
    _ADD_TO_LIST_IF_NOT_PRESENT(_PKG_CONFIG_DEP_NOT_FOR_CONFIG_CMAKE "${PARSED_ARGN_PKG_CONFIG_REQUIRES}")
  ENDIF()
  
  find_package(${PARSED_ARGN_UNPARSED_ARGUMENTS})
  if (${ARGV0}}_FOUND)

    list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_PROJECTS "${ARGV0}")


    string(REPLACE ";" " " PACKAGE_ARGS "${PARSED_ARGN_UNPARSED_ARGUMENTS}")
    list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_FIND_PACKAGE "  find_package(${PACKAGE_ARGS})")
    list(APPEND _PACKAGE_CONFIG_DEPENDENCIES_FIND_DEPENDENCY "  find_dependency(${PACKAGE_ARGS})")

    # Propagate variables changes to the cached values
    set(_PACKAGE_CONFIG_DEPENDENCIES_PROJECTS "${_PACKAGE_CONFIG_DEPENDENCIES_PROJECTS}" CACHE INTERNAL "")
    set(_PACKAGE_CONFIG_DEPENDENCIES_FIND_PACKAGE "${_PACKAGE_CONFIG_DEPENDENCIES_FIND_PACKAGE}" CACHE INTERNAL "")
    set(_PACKAGE_CONFIG_DEPENDENCIES_FIND_DEPENDENCY "${_PACKAGE_CONFIG_DEPENDENCIES_FIND_DEPENDENCY}" CACHE INTERNAL "")

  endif()
ENDMACRO()
```

**Comment by olivier-stasse** [comment]
Thanks for the comments, I will stick to the solution suggested by @gergondet for the optional part in #391

File: package-config.cmake (hunk 1 above)
Comment (olivier-stasse) [reply package-config.cmake:92]: Thanks @gergondet for the feedback. I wanted to update my skills in cmake. 
For the sake of my understanding: 

* if the macro is using caps should we switch to lower case or stay coherent with the macros.

* What is the proper way to deal with arguments in a macro without if/else/endif conditions ? Could you point me towards an example ?

Comment (gergondet) [reply package-config.cmake:92]: > if the macro is using caps should we switch to lower case or stay coherent with the macros.

I think within a single macro we should stay coherent (or use the opportunity to switch all the macro to lower-cases)

Note: I realize it's not the case in this macro but I didn't review that part myself (and we could use some linter/format integration in the process)

> What is the proper way to deal with arguments in a macro without if/else/endif conditions ? Could you point me towards an example ?

In that case, I meant to replace:

```cmake
if(PARSED_ARGN_OPTIONAL OR NOT PARSED_ARGN_REQUIRED)
...
else(PARSED_ARGN_OPTIONAL OR NOT PARSED_ARGN_REQUIRED)
...
endif(PARSED_ARGN_OPTIONAL OR NOT PARSED_ARGN_REQUIRED)
```

with

```cmake
if(PARSED_ARGN_OPTIONAL OR NOT PARSED_ARGN_REQUIRED)
...
else()
...
endif()
```

the repeated conditions have not been necessary for a long time (CMake 2.6.0 see [this answer](https://stackoverflow.com/a/35269674))

File: package-config.cmake (hunk 2 above)
Comment (olivier-stasse) [reply package-config.cmake:105]: Yes we would need to test PARSED_ARGN_REQUIRED to add or not the REQUIRED parameter.

File: package-config.cmake (hunk 1 above)
Comment (olivier-stasse) [reply package-config.cmake:92]: Thanks for the clarification !

**Pull request closed without merging**